### ISSUE TITLE
WITH TOTALS triggers CANNOT_INSERT_NULL_IN_ORDINARY_COLUMN

### ISSUE THREAD

**Issue by qoega** [ClickHouse/ClickHouse#65278]
Minified but slightly another error
```
SELECT
    CAST(c1, 'Int8'),
    anyIf(c2, c3 = '3')
FROM values(('1', '2', '3'))
GROUP BY c1
    WITH TOTALS

Query id: 1f7c23b6-8e72-4156-91aa-b4d0eda26e93

   ┌─CAST(c1, 'Int8')─┬─anyIf(c2, equals(c3, '3'))─┐
1. │                1 │ 2                          │
   └──────────────────┴────────────────────────────┘
↑ Progress: 0.00 rows, 0.00 B (0.00 rows/s., 0.00 B/s.)
1 row in set. Elapsed: 0.002 sec.

Received exception from server (version 24.6.1):
Code: 32. DB::Exception: Received from localhost:9000. DB::Exception: Attempt to read after eof: Cannot parse Int8 from String, because value is too short: while executing 'FUNCTION CAST(c1 :: 0, 'Int8' :: 2) -> CAST(c1, 'Int8') Int8 : 3'. (ATTEMPT_TO_READ_AFTER_EOF)
```

Initially saw in 
```
SELECT query_number::Int8, sumIf(query_duration_ms::UInt64, run_id='0') as cold, sumIf(query_duration_ms::UInt64,run_id='1') as hot, sum(success)/2 as success  FROM file('results.json') group by query_number WITH TOTALS ORDER BY 1 ASC"
....
Code: 349. DB::Exception: Cannot convert NULL value to non-Nullable type: while executing 'FUNCTION CAST(__table1.query_number :: 0, 'Int8'_String :: 1) -> CAST(__table1.query_number, 'Int8'_String) Int8 : 2'. (CANNOT_INSERT_NULL_IN_ORDINARY_COLUMN)
ClickHouse client version 24.6.1.3775 (official build).
```

**Comment by max-vostrikov** [comment]
I do not see any unexpected behaviour here, you are using values() function not in a way you expect - there is no way to check which value is witch column - c1 c2 or c3.

So by changing value function you can get expected result:
[https://fiddle.clickhouse.com/13ef1f21-4b88-4681-ae0a-e875d2a0ece9](https://fiddle.clickhouse.com/13ef1f21-4b88-4681-ae0a-e875d2a0ece9)
```
--returns expected result
SELECT
    CAST(c1, 'Int8'),
    anyIf(c2, c3 = '3')
FROM values('1' as c1, '2' as c2, '3' as c3)
GROUP BY c1
    WITH TOTALS;


--returns all 3 values
SELECT
    CAST(c1, 'Int8')
FROM values('1', '2', '3');


--returns only 1st value - as the value is tuple ('1', '2', '3'). The tuple gets converted to the first element
SELECT
    CAST(c1, 'Int8')
FROM values(('1', '2', '3'));
```

So i do not see anything wrong in the first case.

The error in the second case looks just like a NULL value in your data.
anything I'm missing here?

**Comment by qoega** [comment]
@groeneai can you reproduce?

**Comment by groeneai** [comment]
Yes, both shapes reproduce on today's master (`26.9.1.1`, `88dd163466bc`), on the new and the old analyzer.

The minimal case needs no `values()`:

```sql
SELECT CAST(k, 'Int8'), count()
FROM (SELECT toString(number + 1) AS k FROM numbers(2))
GROUP BY k WITH TOTALS;
-- Code: 32. Attempt to read after eof: Cannot parse Int8 from String, because value is too short
```

The same query without `WITH TOTALS` passes. `WITH ROLLUP` fails the same way.

Mechanism: the extra row has its key columns filled with the type default (documented in `select/group-by.mdx`: "This row will have key columns containing default values (zeros or empty lines)"; the values come from the initial totals state in `TotalsHavingTransform.cpp:111`), and the SELECT-list `CAST` is then evaluated on that synthetic row. A `String` key yields `''` and `Code: 32`; a `Nullable` key yields `NULL` and `Code: 349`. Your two reports are one bug, and only the key's nullability picks the code.

The `file('results.json')` shape reproduces verbatim on a file that contains no `NULL` at all, because JSON inference types `query_number` as `Nullable(Int64)`. With `schema_inference_make_columns_nullable=0` the same query passes and the totals key prints `0`.

So the `st-need-info` premises do not hold:

* `values(('1', '2', '3'))` binds the columns as you assumed: `DESCRIBE` returns `c1`, `c2`, `c3` of type `String`, one row.
* `values('1' AS c1, '2' AS c2, '3' AS c3)` is a different fixture: one column, three rows, with `c1`/`c2`/`c3` in the query resolving to the constant aliases, so the keys are folded constants and the totals row keeps `1`, `2`. It sidesteps the failure rather than contradicting it.
* The `Code: 349` case has no `NULL` in the data (`countIf(query_number IS NULL)` is 0).

`group_by_use_nulls = 1` does not change it, and neither does any `totals_mode`. Workarounds today: group by the cast (`GROUP BY CAST(k, 'Int8')`, whose totals key is `0`), or use `toInt8OrNull`.

If you want this fixed rather than documented, say so and I will pick it up.
